### PR TITLE
adds idle network condition for driver to consider page loaded

### LINKED ISSUE #627: waiting for load event + Xms isn't sufficient time for some sites
one example: https://shop.polymer-project.org. See where the load event fires here:

screen shot 2016-09-01 at 15 29 39

One way this manifests is that the service worker isn't given enough time to even start downloading (so #599 wouldn't help), so the SW and offline tests fail. Presumably the e.g. Speed Index score is also very wrong, as most of the content hasn't actually loaded.

Interestingly, the extension audits the page somewhat successfully because the trace apparently takes a lot longer (a few seconds) to come across to Lighthouse in the extension than on the CLI, giving the page enough time to finish loading the service worker, recording more (most?) of the network requests, etc.

This can be solved for sites specifically instrumenting for LH by #519, but we also need a solution for the general case...something like "we're past load event and the network is calmish".

### PR DESCRIPTION
**Pull request by brendankenny**
fixes #627 

It becomes a little convoluted with the three promises, event listeners, and timeouts, but hopefully this makes it so you can at least reason about them. Happy to entertain other ways to organize.

### CODE REVIEW

**Comment by brendankenny** [comment]
biggest win here are sites that queue stuff up to load after onload, as this now captures their entire load. e.g. from the example in #627, https://shop.polymer-project.org/ gets most of its assets pushed to it after onload, so LH doesn't have much to say about it on master:

screen shot 2016-09-28 at 15 25 25

After this change you get much more meaningful results (ignoring cached `start_url` due to #566 :)
screen shot 2016-09-28 at 15 26 35

**Comment by addyosmani** [comment]
Nice!

The Time to interactive numbers look interesting there. Profiling Shop on an N5+3G we were seeing ~1.75 for it reaching interactivity but this could either be down to 1) testing methodology or 2) this being a single run.

File: lighthouse-core/lib/network-recorder.js [hunk 1]
```diff
@@ -26,10 +27,36 @@ class NetworkRecorder extends EventEmitter {
...
 
-    // TODO(bckenny): loadingFailed calls are not recorded in REQUEST_FINISHED.
+    this.startedRequestCount = 0;
+    this.finishedRequestCount = 0;
+    this.idleState = true;
+
+    this.networkManager.addEventListener(this.EventTypes.RequestStarted, _ => {
+      this.startedRequestCount++;
+
+      const activeCount = this.startedRequestCount - this.finishedRequestCount;
```
Comment (paulirish) [inline lighthouse-core/lib/network-recorder.js]: move to a method that `isIdle()` sits on top of?

File: lighthouse-core/gather/drivers/driver.js [hunk 2]
```diff
@@ -263,50 +263,148 @@ class Driver {
...
+   * Returns a promise that resolves when:
+   * - it's been pauseAfterLoadMs milliseconds after both onload and the network
+   * has gone idle, or
+   * - MAX_WAIT_FOR_LOAD_EVENT milliseconds have passed.
+   * See https://github.com/GoogleChrome/lighthouse/issues/627 for more.
+   * @param {number} pauseAfterLoadMs
+   * @return {!Promise}
+   * @private
+   */
+  _waitForLoad(pauseAfterLoadMs) {
```
Comment (brendankenny) [inline lighthouse-core/gather/drivers/driver.js]: sounds good to me

File: lighthouse-core/lib/network-recorder.js [hunk 3]
```diff
@@ -26,10 +27,36 @@ class NetworkRecorder extends EventEmitter {
...
+          ` (${this.startedRequestCount} started and ${this.finishedRequestCount} finished).`);
+
+      if (this.idleState) {
+        this.idleState = false;
+        this.emit('networkbusy');
+      }
+    });
+
     this.networkManager.addEventListener(this.EventTypes.RequestFinished, request => {
+      this.finishedRequestCount++;
```
Comment (brendankenny) [inline lighthouse-core/lib/network-recorder.js]: > what about failed requests?

as discussed, `RequestFinished` includes failed requests

> and should we handle websocket connections specially or just allow the timeout to address em ?

I actually haven't had any issues with sites that use web sockets so far, weirdly enough. Any good test sites for this?

**Comment by samccone** [comment]
I wonder if instead of this approach we should allow users to emit a mark event to tell lighthouse to stop.
I worry that we are continuing to add more and more methods to determine when a page is done, however who knows better than the actual page.

**Comment by brendankenny** [comment]
This could definitely become a slippery slope of accumulating heuristics, and we should make sure that it doesn't. We should definitely do the #519 thing and have a way for authors to mark a page as "done", the problem is all the sites that won't be instrumented also need some kind of reasonable way of guessing when they're finished.

@paulirish suggested the network heuristic in #627 after looking into how to make the polymer shop example (and future PRPL sites) work with Lighthouse, and we found out shortly later that the folks doing FMP in Chrome are using something very similar when `onload` isn't useful for whatever reason and have found it very useful. @paulirish might be able to comment more on that.

File: lighthouse-core/gather/drivers/driver.js [hunk 4]
```diff
@@ -263,50 +263,148 @@ class Driver {
...
+      return function() {
+        log.verbose('Driver', 'loadEventFired and network considered idle');
+        clearTimeout(maxTimeoutHandle);
+      };
+    });
+
+    // Last resort timeout. Resolves MAX_WAIT_FOR_LOAD_EVENT ms from now on
+    // cleanup function that removes loadEvent and network idle listeners.
+    const maxTimeoutPromise = new Promise((resolve, reject) => {
+      maxTimeoutHandle = setTimeout(resolve, MAX_WAIT_FOR_LOAD_EVENT);
```
Comment (brendankenny) [inline lighthouse-core/gather/drivers/driver.js]: done

File: lighthouse-core/lib/network-recorder.js [hunk 5]
```diff
@@ -26,10 +27,36 @@ class NetworkRecorder extends EventEmitter {
     this._records = recordArray;
     this.networkManager = NetworkManager.createWithFakeTarget();
 
-    // TODO(bckenny): loadingFailed calls are not recorded in REQUEST_FINISHED.
+    this.startedRequestCount = 0;
+    this.finishedRequestCount = 0;
+    this.idleState = true;
```
Comment (brendankenny) [inline lighthouse-core/lib/network-recorder.js]: they are different to keep track of which event we've actually emitted, but good point, we can use the actual number of active requests to derive that state

File: lighthouse-core/lib/network-recorder.js (hunk 1 above)
Comment (brendankenny) [inline lighthouse-core/lib/network-recorder.js]: done

File: lighthouse-core/lib/network-recorder.js [hunk 6]
```diff
@@ -18,6 +18,7 @@
 
 const NetworkManager = require('./web-inspector').NetworkManager;
 const EventEmitter = require('events').EventEmitter;
+const log = require('../lib/log.js');
 
 class NetworkRecorder extends EventEmitter {
   constructor(recordArray) {
```
Comment (brendankenny) [inline lighthouse-core/lib/network-recorder.js]: boom

**Comment by brendankenny** [comment]
PTAL

**Pull request merged**